File: src/main/java/com/example/SnowflakeConfig.java
package com.example;

/**
 * Configuration class for Snowflake connection
 */
public class SnowflakeConfig {
    private String url;
    private String user;
    private String password;
    private String database;
    private String schema;
    private String role;
    private String warehouse;

    public SnowflakeConfig() {}

    public SnowflakeConfig(String url, String user, String password, String database, 
                          String schema, String role, String warehouse) {
        this.url = url;
        this.user = user;
        this.password = password;
        this.database = database;
        this.schema = schema;
        this.role = role;
        this.warehouse = warehouse;
    }

    // Getters and Setters
    public String getUrl() {
        return url;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public String getUser() {
        return user;
    }

    public void setUser(String user) {
        this.user = user;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public String getDatabase() {
        return database;
    }

    public void setDatabase(String database) {
        this.database = database;
    }

    public String getSchema() {
        return schema;
    }

    public void setSchema(String schema) {
        this.schema = schema;
    }

    public String getRole() {
        return role;
    }

    public void setRole(String role) {
        this.role = role;
    }

    public String getWarehouse() {
        return warehouse;
    }

    public void setWarehouse(String warehouse) {
        this.warehouse = warehouse;
    }

    @Override
    public String toString() {
        return "SnowflakeConfig{" +
                "url='" + url + '\'' +
                ", user='" + user + '\'' +
                ", password='[HIDDEN]'" +
                ", database='" + database + '\'' +
                ", schema='" + schema + '\'' +
                ", role='" + role + '\'' +
                ", warehouse='" + warehouse + '\'' +
                '}';
    }
}

File: src/main/scala/com/example/SnowflakeConfig.scala
package com.example

/**
 * Scala case class for Snowflake connection configuration
 * Immutable data structure with default values
 */
case class SnowflakeConfig(
  url: String,
  user: String,
  password: String,
  database: String,
  schema: String,
  role: String,
  warehouse: String
) {
  
  /**
   * Create a copy with updated password (for security)
   */
  def withPassword(newPassword: String): SnowflakeConfig = 
    this.copy(password = newPassword)
  
  /**
   * Create a copy with hidden password for logging
   */
  def forLogging: SnowflakeConfig = 
    this.copy(password = "[HIDDEN]")
  
  /**
   * Validate configuration
   */
  def validate: Either[String, SnowflakeConfig] = {
    val errors = scala.collection.mutable.ListBuffer[String]()
    
    if (url.isEmpty) errors += "URL cannot be empty"
    if (user.isEmpty) errors += "User cannot be empty"
    if (password.isEmpty) errors += "Password cannot be empty"
    if (database.isEmpty) errors += "Database cannot be empty"
    if (schema.isEmpty) errors += "Schema cannot be empty"
    if (role.isEmpty) errors += "Role cannot be empty"
    if (warehouse.isEmpty) errors += "Warehouse cannot be empty"
    
    if (errors.nonEmpty) {
      Left(errors.mkString(", "))
    } else {
      Right(this)
    }
  }
  
  /**
   * Get JDBC URL with parameters
   */
  def getJdbcUrl: String = {
    if (url.contains("?")) {
      url
    } else {
      s"$url?db=$database&schema=$schema&warehouse=$warehouse&role=$role"
    }
  }
  
  override def toString: String = {
    s"SnowflakeConfig(url='$url', user='$user', password='[HIDDEN]', " +
    s"database='$database', schema='$schema', role='$role', warehouse='$warehouse')"
  }
}

object SnowflakeConfig {
  
  /**
   * Create default configuration
   */
  def default: SnowflakeConfig = SnowflakeConfig(
    url = "jdbc:snowflake://a_icg_dev.us-east-2.aws.snowflakecomputing.com",
    user = "F_ICG_DEV_177688_ALERTS",
    password = "YOUR_PASSWORD_HERE",
    database = "D_ICG_DEV_177688_MASTER",
    schema = "LANDING",
    role = "R_ICG_DEV_177688_APPADMIN",
    warehouse = "W_ICG_DEV_177688_DEFAULT_XS"
  )
  
  /**
   * Create configuration from individual parameters
   */
  def apply(url: String, user: String, password: String, database: String, 
           schema: String, role: String, warehouse: String): SnowflakeConfig = {
    new SnowflakeConfig(url, user, password, database, schema, role, warehouse)
  }
}
